{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "98afb9b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e940e7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7859994, 43)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"../data/variant_summary.txt\", sep = \"\\t\", low_memory = False)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5d4b968b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OriginSimple\n",
       "germline            7683295\n",
       "unknown              142500\n",
       "somatic               12007\n",
       "not applicable         9241\n",
       "not provided           7702\n",
       "germline/somatic       5249\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['OriginSimple'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dc283fdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kept features with count of missing\n",
      "11015\n",
      "102181\n",
      "102181\n"
     ]
    }
   ],
   "source": [
    "print('Kept features with count of missing')\n",
    "print(df['Cytogenetic'].value_counts().get('-'))\n",
    "print(df['ReferenceAlleleVCF'].value_counts().get('na'))\n",
    "print(df['AlternateAlleleVCF'].value_counts().get('na'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "781ffd7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "useless_columns = ['#AlleleID', 'GeneID', 'Name', 'HGNC_ID', 'RS# (dbSNP)', 'nsv/esv (dbVar)', 'RCVaccession',\n",
    "                   'PhenotypeIDS', 'OriginSimple', 'ReferenceAllele', 'AlternateAllele', 'Guidelines', 'OtherIDs',\n",
    "                   'VariationID', 'SomaticClinicalImpactLastEvaluated', 'SomaticClinicalImpact', 'Oncogenicity', \n",
    "                   'OncogenicityLastEvaluated', 'ReviewStatusOncogenicity', 'SCVsForAggregateGermlineClassification', 'ReviewStatus',\n",
    "                   'SCVsForAggregateGermlineClassification', 'SCVsForAggregateSomaticClinicalImpact', 'SCVsForAggregateOncogenicityClassification', 'SubmitterCategories',\n",
    "                   'ClinicalSignificance', 'LastEvaluated', 'PhenotypeList', 'Assembly', 'NumberSubmitters', 'TestedInGTR', 'ReviewStatusClinicalImpact', 'Origin',\n",
    "                   'ChromosomeAccession', 'PositionVCF'\n",
    "                   ]\n",
    "features = df.drop(columns = useless_columns)\n",
    "features = features.loc[features['ClinSigSimple'] != -1]\n",
    "features = features.replace(['na', '-', -1], np.nan)\n",
    "column_names = list(features.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a040db7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Type', 'GeneSymbol', 'ClinSigSimple', 'Chromosome', 'Start', 'Stop',\n",
      "       'Cytogenetic', 'PositionVCF', 'ReferenceAlleleVCF',\n",
      "       'AlternateAlleleVCF'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(features.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9ee690bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cytogenetic\n",
       "16p13.3        121161\n",
       "2q31.2          78944\n",
       "19p13.2         70471\n",
       "19p13.3         65063\n",
       "19q13.2         62636\n",
       "                ...  \n",
       "18q12.2-23          1\n",
       "2p23-p22            1\n",
       "8q12-q13            1\n",
       "1p31.1-22.1         1\n",
       "5q33.1-33.3         1\n",
       "Name: count, Length: 3816, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features['Cytogenetic'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4126c0c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlternateAlleleVCF\n",
      "T                                                               2224657\n",
      "A                                                               2215020\n",
      "C                                                               1559133\n",
      "G                                                               1541723\n",
      "CT                                                                12757\n",
      "                                                                 ...   \n",
      "GGGGCG                                                                1\n",
      "GCCGCTGCCGCTGCCGCTGCC                                                 1\n",
      "GTGAAAGCCTCCATCGTTTCATGCTCTCTGCCCTTCGGCCCTGAAAGCCTCCATCGTTTC          1\n",
      "TCCTTTCCTCCTCCTCCTC                                                   1\n",
      "GCAGGACAGCGTCCCCAGTGAGTGCACCCA                                        1\n",
      "Name: count, Length: 19572, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(features['AlternateAlleleVCF'].value_counts())\n",
    "to_write = features['AlternateAlleleVCF'].value_counts()\n",
    "\n",
    "with open(\"view_full_output.txt\", \"w\") as f:\n",
    "    f.write(to_write.to_string())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene-variation-effects",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
