{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b49b4c0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyliftover import LiftOver\n",
    "from myvariant import MyVariantInfo\n",
    "import numpy as np\n",
    "\n",
    "def convert_row_to_hgvs(row) -> str:\n",
    "    start = row['Start'].astype(int)\n",
    "\n",
    "    if row['Type'] == \"single nucleotide variant\":\n",
    "        return f\"chr13:g.{start}{row['ReferenceAlleleVCF']}>{row['AlternateAlleleVCF']}\"\n",
    "    if row['Type'] == \"Deletion\":\n",
    "        return f\"chr13:g.{start}_{row['Stop']}del\"\n",
    "\n",
    "    type_mapping = {\n",
    "        \"Insertion\": \"ins\",\n",
    "        \"Duplication\": \"dup\",\n",
    "        \"Indel\": \"delins\",\n",
    "    }\n",
    "\n",
    "    return f\"chr13:g.{start}_{row['Stop']}{type_mapping[row['Type']]}{row['ReferenceAlleleVCF']}\"\n",
    "\n",
    "\n",
    "    # # Get allele frequency\n",
    "    # lo = LiftOver('hg38', 'hg19')\n",
    "    # df['Start'] = np.where(df['Start'] >= gene_start_position_hg38, [lo.convert_coordinate('chr17', start)[0][1] for start in df['Start']], df['Start'])\n",
    "    # df['Stop'] = np.where(df['Stop'] >= gene_start_position_hg38, [lo.convert_coordinate('chr17', start)[0][1] for start in df['Stop']], df['Stop'])\n",
    "\n",
    "    # mv_info = MyVariantInfo()\n",
    "    # df['Frequency'] = [mv_info.getvariant(convert_row_to_hgvs(row)) for _, row in df.iterrows()]"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
